tools/@aws-cdk/spec2cdk/lib/cdk/resource-decider.ts

```diff
@@ -58,39 +61,73 @@ export class ResourceDecider {
     for (const arn of possibleArnNames) {
-      const att = this.classAttributeProperties.filter((a) => a.propertySpec.name === attributePropertyName(arn));
-      const prop = this.propsProperties.filter((p) => p.propertySpec.name === propertyNameFromCloudFormation(arn));
-      if (att.length > 0 || prop.length > 0) {
-        return att[0] ? att[0].propertySpec : prop[0].propertySpec;
-      }
+      const att = this.classAttributeProperties.find((a) => a.propertySpec.name === attributePropertyName(arn));
+      if (att) { return att.propertySpec; }
+      // TODO: double check -- if an arn is a property, not an attribute, and not a primaryIdentifier
+      // it is not getting returned by cloudformation in any way, right?
+      // const prop = this.propsProperties.find((p) => p.propertySpec.name === propertyNameFromCloudFormation(arn));
+      // if (prop) {
+      //   return this.convertPropertySpecToRefAttribute(prop.propertySpec);
+      // }
     }
     return;
   }
 
+  private convertPropertySpecToRefAttribute(propSpec: PropertySpec): PropertySpec {
+    return {
+      ...propSpec,
+      name: attributePropertyName(propSpec.name[0].toUpperCase() + propSpec.name.slice(1)),
+      docs: {
+        ...propSpec.docs,
+        remarks: propSpec.docs?.remarks?.concat(['\n', '@cloudformationAttribute Ref'].join('\n')),
+      },
+      immutable: true,
+    };
+  }
+
   private convertPrimaryIdentifier() {
-    for (const cfnName of this.resource.primaryIdentifier ?? []) {
+    if (this.resource.primaryIdentifier === undefined) { return; }
+    for (let i = 0; i < (this.resource.primaryIdentifier).length; i++) {
+      const cfnName = this.resource.primaryIdentifier[i];
       const att = this.findAttributeByName(attributePropertyName(cfnName));
       const prop = this.findPropertyByName(propertyNameFromCloudFormation(cfnName));
       if (att) {
         this.primaryIdentifier.push(att);
       } else if (prop) {
-        // rename the prop name as an attribute name, since it is gettable by ref
-        this.primaryIdentifier.push({
-          ...prop,
-          name: attributePropertyName(prop.name[0].toUpperCase() + prop.name.slice(1)),
-          docs: {
-            ...prop.docs,
-            remarks: prop.docs?.remarks?.concat(['\n', `@cloudformationRef ${prop.name}`].join('\n')),
-          },
+        const propSpec = prop.propertySpec;
+
+        // Build an attribute out of the property we're getting
+        // Create initializer for new attribute, if possible
+        let initializer: Expression | undefined = undefined;
+        if (propSpec.type === Type.STRING) { // handling only this case for now
+          if (this.resource.primaryIdentifier!.length === 1) {
+            initializer = CDK_CORE.tokenAsString($this.ref);
+          } else {
+            initializer = CDK_CORE.tokenAsString($Fn.select(expr.directCode(i.toString()), $Fn.split(expr.directCode('\'|\''), $this.ref)));
```

Comment:
Please avoid `expr.directCode()` as much as possible.

```suggestion
            initializer = CDK_CORE.tokenAsString($Fn.select(expr.lit(i), $Fn.split(expr.lit('|'), $this.ref)));
```